File: apps/server/models/bill.py
from extensions import db
from datetime import datetime

class BillCategory(db.Model):
    __tablename__ = 'app_bill_category'
    
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(50), nullable=False)
    parent_id = db.Column(db.Integer, db.ForeignKey('app_bill_category.id'))
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # 自关联
    children = db.relationship('BillCategory', backref=db.backref('parent', remote_side=[id]))

class Bill(db.Model):
    __tablename__ = 'app_bill'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('sys_user.id'), nullable=False)
    category_id = db.Column(db.Integer, db.ForeignKey('app_bill_category.id'), nullable=False)
    amount = db.Column(db.Float, nullable=False)
    order_name = db.Column(db.String(200))
    spending_time = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    is_deleted = db.Column(db.Boolean, default=False)

    # 关系
    category = db.relationship('BillCategory', backref=db.backref('bills', lazy=True))

    def to_dict(self):
        return {
            'id': self.id,
            'user_id': self.user_id,
            'category_id': self.category_id,
            'amount': self.amount,
            'order_name': self.order_name,
            'spending_time': self.spending_time.strftime('%Y-%m-%d %H:%M:%S'),
            'created_at': self.created_at.strftime('%Y-%m-%d %H:%M:%S'),
            'updated_at': self.updated_at.strftime('%Y-%m-%d %H:%M:%S'),
            'is_deleted': self.is_deleted
        } 

File: apps/server/models/Bill.js
const mongoose = require('mongoose');

const billCategorySchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
    maxlength: 50
  },
  parentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'BillCategory',
    default: null
  },
  icon: {
    type: String,
    default: ''
  },
  color: {
    type: String,
    default: '#409EFF'
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true
});

const billSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  categoryId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'BillCategory',
    required: true
  },
  amount: {
    type: Number,
    required: true,
    min: 0
  },
  orderName: {
    type: String,
    required: true,
    trim: true,
    maxlength: 200
  },
  description: {
    type: String,
    trim: true,
    maxlength: 500,
    default: ''
  },
  spendingTime: {
    type: Date,
    required: true,
    default: Date.now
  },
  tags: [{
    type: String,
    trim: true,
    maxlength: 20
  }],
  isDeleted: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Indexes for better query performance
billSchema.index({ userId: 1, isDeleted: 1 });
billSchema.index({ userId: 1, spendingTime: -1 });
billSchema.index({ categoryId: 1 });

// Populate category information when querying
billSchema.pre(/^find/, function(next) {
  this.populate({
    path: 'categoryId',
    select: 'name icon color'
  });
  next();
});

// Static method to get user's spending statistics
billSchema.statics.getSpendingStats = async function(userId, startDate, endDate) {
  const pipeline = [
    {
      $match: {
        userId: new mongoose.Types.ObjectId(userId),
        isDeleted: false,
        spendingTime: {
          $gte: startDate,
          $lte: endDate
        }
      }
    },
    {
      $group: {
        _id: '$categoryId',
        totalAmount: { $sum: '$amount' },
        count: { $sum: 1 },
        avgAmount: { $avg: '$amount' }
      }
    },
    {
      $lookup: {
        from: 'billcategories',
        localField: '_id',
        foreignField: '_id',
        as: 'category'
      }
    },
    {
      $unwind: '$category'
    },
    {
      $project: {
        categoryName: '$category.name',
        categoryIcon: '$category.icon',
        categoryColor: '$category.color',
        totalAmount: 1,
        count: 1,
        avgAmount: { $round: ['$avgAmount', 2] }
      }
    },
    {
      $sort: { totalAmount: -1 }
    }
  ];

  return this.aggregate(pipeline);
};

// Soft delete method
billSchema.methods.softDelete = function() {
  this.isDeleted = true;
  return this.save();
};

const BillCategory = mongoose.model('BillCategory', billCategorySchema);
const Bill = mongoose.model('Bill', billSchema);

module.exports = { Bill, BillCategory };
